lib/statsd/instrument/dogstatsd_datagram_builder.rb

```diff
@@ -8,4 +8,64 @@ class StatsD::Instrument::DogStatsDDatagramBuilder < StatsD::Instrument::Datagra
   def latency_metric_type
     :d
   end
+
+  # Constricts an event datagram.
+  #
+  # @param [String] title Event title.
+  # @param [String] text Event description. Newlines are allowed.
+  # @param [Time] timestamp The of the event. If not provided,
+  #   Datadog will interpret it as the current timestamp.
+  # @param [String] hostname A hostname to associate with the event.
+  # @param [String] aggregation_key An aggregation key to group events with the same key.
+  # @param [String] priority Priority of the event. Either "normal" (default) or "low".
+  # @param [String] source_type_name The source type of the event.
+  # @param [String] alert_type Either "error", "warning", "info" (default) or "success".
+  # @param [Array, Hash] tags Tags to associate with the event.
+  # @return [String] The correctly formatted service check datagram
+  #
+  # @see https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/#events
+  def _e(title, text, timestamp: nil, hostname: nil, aggregation_key: nil, priority: nil,
+    source_type_name: nil, alert_type: nil, tags: nil)
+
+    escaped_title = title.gsub("\n", '\n')
+    escaped_text = text.gsub("\n", '\n')
+    tags = normalize_tags(tags) + default_tags
+
+    datagram = +"_e{#{escaped_title.length},#{escaped_text.length}}:#{escaped_title}|#{escaped_text}"
+    datagram << "|d:#{timestamp.to_i}" if timestamp
+    datagram << "|h:#{hostname}" if hostname
+    datagram << "|k:#{aggregation_key}" if aggregation_key
+    datagram << "|p:#{priority}" if priority
+    datagram << "|s:#{source_type_name}" if source_type_name
+    datagram << "|t:#{alert_type}" if alert_type
+    datagram << "|##{tags.join(',')}" unless tags.empty?
+    datagram
+  end
+
+  # Constricts a service check datagram.
+  #
+  # @param [String] name Name of the service
+  # @param [Symbol] status Either `:ok`, `:warning`, `:critical` or `:unknown`
+  # @param [Time] timestamp The moment when the service was checked. If not provided,
+  #   Datadog will interpret it as the current timestamp.
+  # @param [String] hostname A hostname to associate with the check.
+  # @param [Array, Hash] tags Tags to associate with the check.
+  # @param [String] message A message describing the current state of the service check.
+  # @return [String] The correctly formatted service check datagram
+  #
+  # @see https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/#service-checks
+  def _sc(name, status, timestamp: nil, hostname: nil, tags: nil, message: nil)
+    status_number = SERVICE_CHECK_STATUS_VALUES.fetch(status)
+    tags = normalize_tags(tags) + default_tags
+
+    datagram = +"_sc|#{normalize_name(name)}|#{status_number}"
+    datagram << "|d:#{timestamp.to_i}" if timestamp
+    datagram << "|h:#{hostname}" if hostname
+    datagram << "|##{tags.join(',')}" unless tags.empty?
```

Comment:
I'm not sure if the order matters to datadog, but this was last in `StatsD::Instrument::Backends::UDPBackend::DogStatsDProtocol`, so probably should be here so that a DogStatsDDatagramCompatibilityTest test could be written for it.

Reply:
Interesting. This is indeed a difference. The DogstatsD documentation is quite explicit that `m` has to be the final parameter though:

> Add a message describing the current state of the service check. This field MUST be positioned last among the metadata fields. No default.

I have confirmed with the compatibility test that the behavior is actually different, but I am pretty sure this is a bug in the old implementation.

Reply:
:ok:  we might not have actually used the tags with this

Reply:
#198 and #199 will fix some issues with the current implementation,, after which I can add a compatibility test.